modules/fee-market/src/s2s/payment.rs

```diff
@@ -139,70 +139,98 @@ where
-					reward_item.to_treasury = Some(message_fee.saturating_sub(base_fee));
-
-					// AssignedRelayersRewardRatio * base fee => slot relayer
-					let slot_relayer_reward = T::AssignedRelayersRewardRatio::get() * base_fee;
-					reward_item.to_slot_relayer = Some((who, slot_relayer_reward));
-
-					let bridger_relayers_reward = base_fee.saturating_sub(slot_relayer_reward);
-					// MessageRelayersRewardRatio * (1 - AssignedRelayersRewardRatio) * base_fee =>
-					// message relayer
-					message_reward = T::MessageRelayersRewardRatio::get() * bridger_relayers_reward;
-					// ConfirmRelayersRewardRatio * (1 - AssignedRelayersRewardRatio) * base_fee =>
-					// confirm relayer
-					confirm_reward = T::ConfirmRelayersRewardRatio::get() * bridger_relayers_reward;
-				} else {
-					// The order delivery is delay, slash occurs.
-					let mut total_slash = message_fee;
-
-					// calculate slash amount
-					let mut amount: BalanceOf<T, I> = T::Slasher::slash(
-						order.locked_collateral,
-						order.delivery_delay().unwrap_or_default(),
-					);
-					if let Some(slash_protect) = Pallet::<T, I>::collateral_slash_protect() {
-						amount = sp_std::cmp::min(amount, slash_protect);
-					}
-
-					// Slash order's assigned relayers
-					let mut assigned_relayers_slash = BalanceOf::<T, I>::zero();
-					for assigned_relayer in order.relayers_slice() {
-						let report = SlashReport::new(&order, assigned_relayer.id.clone(), amount);
-						let slashed = do_slash::<T, I>(
-							&assigned_relayer.id,
-							relayer_fund_account,
-							amount,
-							report,
+				match order.delivery_info(order_confirm_time) {
+					// The order has been confirmed at the first assigned relayers slot, no slash
+					// happens.
+					Some((relayer_index, slot_relayer_id, base_fee)) if relayer_index == 0 => {
+						cal_rewards_before_deadline::<T, I>(
+							&slot_relayer_id,
+							&entry.relayer,
+							&confirm_relayer,
+							total_reward,
+							base_fee,
+							&mut reward_item,
 						);
-						assigned_relayers_slash += slashed;
-					}
-					total_slash += assigned_relayers_slash;
-
-					// MessageRelayersRewardRatio total slash => message relayer
-					message_reward = T::MessageRelayersRewardRatio::get() * total_slash;
-					// ConfirmRelayersRewardRatio total slash => confirm relayer
-					confirm_reward = T::ConfirmRelayersRewardRatio::get() * total_slash;
+					},
+					// The order is confirmed after the first slot and the assigned relayers
+					// corresponding to the previous slots are penalised.
+					Some((relayer_index, slot_relayer_id, base_fee)) if relayer_index >= 1 => {
+						// Calculate the penalty for the previous relayers.
+						let lazy_relayers = order
+							.relayers_slice()
+							.into_iter()
+							.take(relayer_index)
+							.collect::<Vec<_>>();
+
+						for relayer in lazy_relayers {
+							let amount = slash_assigned_relayer::<T, I>(
+								&order,
+								&relayer.id,
+								relayer_fund_account,
```

Comment:
What is `relayer_fund_account`?

Request docs in the function.

Reply:
Okay

Reply:
done